### DISCUSSION TITLE
hints as to what made a boolean expression fail

### DISCUSSION THREAD

**Discussion by crystalrage777** [sympy/sympy#24299] [Q&A] [answered]
Hi,

I have Boolean expressions that say if something should or should not run, these expressions are nested and sometimes I'm interested at what made them fail, and I was wondering if there was an existing solution for it.

Say I have the Boolean expression:
a & b & (c | d) & (!a | (e & f))

If this fails, for the sake of inspection I might be interested at the reason, for example if a was false or b, or both.

The ad-hoc solution I was thinking about was to name groups and then be able to give a path to what failed, for example:
the previous expression turns into
(:case-1 a) & (:case-2 b) & (:case3 (:case3a c) | (:case3b d) & (:case4 (:case4a !a) | (:case4b (:case4b1 e) & (:case4b2 f)))

using this I could say that the reason the expression failed was because "case4" was wrong because both "case4a" and "case4b" were wrong etc.

Is this a known problem, I wasn't able to find anything about it..

THANKS :)

**Comment by asmeurer** [comment]
What you're looking for are the models of `NOT expr`, which can be found with `satisfiable`:

```
>>> expr = a & b & (c | d) & (~a | (e & f))
>>> satisfiable(~expr)
{a: False, b: False, c: False, d: False, e: False, f: False}
```

You can get all possibilities with `all_models=True` but be aware that in general there are exponentially many.

**Comment by oscarbenjamin** [comment, marked as answer]
You omit some of the context for what you are doing here. I presume that you are substituting values for `a`, `b`, ... and if the result comes out to be True or False then you would like to know which subexpressions contributed to that. I would make a recursive function that checks the type of the expression like:
```python
from sympy import *

def investigate(expr, values, result=False, _recurse=0):
    """Answer why expr.subs(values) gives result"""
    assert expr.subs(values) == result

    indent = _recurse * ' '
    print(indent, expr, '->', expr.subs(values))

    if (type(expr), result) in [(And, False), (Or, True)]:
        # Show only clauses that make the And False or the Or True
        for clause in expr.args:
            if clause.subs(values) == result:
                investigate(clause, values, result, _recurse+1)
    elif type(expr) in (And, Or):
        # All clauses are needed to make the And True or the Or False.
        for clause in expr.args:
            investigate(clause, values, result, _recurse+1)
    elif type(expr) is Not:
        investigate(expr.args[0], values, not result, _recurse+1)
    elif type(expr) is Symbol:
        print("root cause:", expr, '=', result)
    else:
        raise NotImplementedError

a, b, c, d, e, f = symbols('a, b, c, d, e, f')
expr = a & b & (c | d) & (~a | (e & f))
values = {a:True, b:False, c:True, d:True, e:True, f:False}
investigate(expr, values)
```
This outputs:
```
 a & b & (c | d) & (~a | (e & f)) -> False
  b -> False
root cause: b = False
  ~a | (e & f) -> False
   ~a -> False
    a -> True
root cause: a = True
   e & f -> False
    f -> False
root cause: f = False
```
It's not trivial to define exactly what you want here. For example do you want sufficient or necessary conditions to make the expression false? Here we can see that `b = False` is sufficient to make the whole expression False.  On the other hand neither `a = True` or `f = False` is sufficient in isolation but together they falsify a whole clause of the `And` which is sufficient to make `expr` False.

**Reply by crystalrage777** [reply]
Yes, this is what I meant.

Thanks!
